File: packages/server-sdk/src/resources/live-streams.ts
import type { RavenHttpClient } from '../http-client';
import type {
  AddHostParams,
  CreateLiveStreamParams,
  IssuedStreamCredential,
  LiveStream,
  ListLiveStreamsParams,
  UpdateLiveStreamParams,
} from '../types';

/**
 * Livqeno Live Streaming, server-side (Phase 14).
 *
 * A stream composes an RTC room and a chat conversation. This resource only
 * owns lifecycle and role bookkeeping on top of those, following the same
 * reuse discipline the backend itself does.
 *
 * `addHost` and `createViewerToken` are the security-critical pair. The role
 * your caller ends up with is decided entirely by which method you call,
 * never by a field in the request body. A viewer token is always
 * subscribe-only, and there's no way to ask this SDK for anything else.
 * `addHost`, called from your own backend after your own auth check, is the
 * only route to publish access.
 */
export class LiveStreamsResource {
  constructor(private readonly http: RavenHttpClient) {}

  create(params: CreateLiveStreamParams): Promise<LiveStream> {
    return this.http.request<LiveStream>('/v1/live-streams', { method: 'POST', body: params });
  }

  list(params: ListLiveStreamsParams = {}): Promise<LiveStream[]> {
    const query = params.status ? `?status=${encodeURIComponent(params.status)}` : '';
    return this.http.request<LiveStream[]>(`/v1/live-streams${query}`);
  }

  /** Includes the stream's live viewer count. `list()` doesn't, to avoid an SFU round trip per row. */
  get(streamId: string): Promise<LiveStream> {
    return this.http.request<LiveStream>(`/v1/live-streams/${encodeURIComponent(streamId)}`);
  }

  update(streamId: string, params: UpdateLiveStreamParams): Promise<LiveStream> {
    return this.http.request<LiveStream>(`/v1/live-streams/${encodeURIComponent(streamId)}`, {
      method: 'PATCH',
      body: params,
    });
  }

  /** CREATED → LIVE. */
  start(streamId: string): Promise<LiveStream> {
    return this.http.request<LiveStream>(`/v1/live-streams/${encodeURIComponent(streamId)}/start`, {
      method: 'POST',
    });
  }

  /** LIVE → ENDED, and that's terminal. You can't restart an ended stream; create a new one. */
  end(streamId: string): Promise<LiveStream> {
    return this.http.request<LiveStream>(`/v1/live-streams/${encodeURIComponent(streamId)}/end`, {
      method: 'POST',
    });
  }

  /**
   * Registers a host or co-host and mints full-publish RTC credentials plus
   * moderator-or-above chat credentials, all in one call. Call it again with
   * the same identity to re-mint fresh ones, say after the originals
   * expired.
   */
  addHost(streamId: string, params: AddHostParams): Promise<IssuedStreamCredential> {
    return this.http.request<IssuedStreamCredential>(`/v1/live-streams/${encodeURIComponent(streamId)}/hosts`, {
      method: 'POST',
      body: params,
    });
  }

  /** A soft removal. The host's chat history in the stream survives. */
  removeHost(streamId: string, identity: string): Promise<void> {
    return this.http.request<void>(
      `/v1/live-streams/${encodeURIComponent(streamId)}/hosts/${encodeURIComponent(identity)}`,
      { method: 'DELETE' },
    );
  }

  /** Always subscribe-only on RTC and MEMBER on chat. See the class doc. */
  createViewerToken(streamId: string, identity: string): Promise<IssuedStreamCredential> {
    return this.http.request<IssuedStreamCredential>(`/v1/live-streams/${encodeURIComponent(streamId)}/viewer-tokens`, {
      method: 'POST',
      body: { identity },
    });
  }

  /**
   * A clean-leave signal for `live_stream.viewer_left`. Not a disconnect
   * detector: Livqeno has no way to spot an abrupt viewer disconnect in this
   * phase. Call it when your own app knows a viewer left.
   */
  leave(streamId: string, identity: string): Promise<void> {
    return this.http.request<void>(`/v1/live-streams/${encodeURIComponent(streamId)}/leave`, {
      method: 'POST',
      body: { identity },
    });
  }
}

File: sdks/python/src/raven/resources/live_streams.py
"""Raven Live Streaming, server-side (Phase 14). A stream composes an RTC
room and a chat conversation — this resource only owns lifecycle and role
bookkeeping on top of them, the same reuse discipline the backend itself
follows.

``add_host()``/``create_viewer_token()`` are the security-critical methods:
the role your caller ends up with is entirely determined by which method
you call, never by a field the request body accepts. A viewer token is
always subscribe-only; there is no way to ask this SDK for anything else —
``add_host()`` (called from your own backend, after your own auth check) is
the only path to publish access.
"""

from __future__ import annotations

from typing import Any, cast

from .._async_http import AsyncRavenHttpClient
from .._http import RavenHttpClient
from .._types import (
    AddHostParams,
    CreateLiveStreamParams,
    IssuedStreamCredential,
    LiveStream,
    LiveStreamStatus,
    UpdateLiveStreamParams,
)


def _drop_none(payload: dict[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in payload.items() if value is not None}


def _create_body(params: CreateLiveStreamParams) -> dict[str, Any]:
    return _drop_none(
        {
            "title": params.title,
            "hostIdentity": params.host_identity,
            "description": params.description,
            "thumbnailUrl": params.thumbnail_url,
            "category": params.category,
            "tags": params.tags,
            "language": params.language,
            "visibility": params.visibility,
            "metadata": params.metadata,
            "scheduledAt": params.scheduled_at,
        }
    )


def _update_body(params: UpdateLiveStreamParams) -> dict[str, Any]:
    return _drop_none(
        {
            "title": params.title,
            "description": params.description,
            "thumbnailUrl": params.thumbnail_url,
            "category": params.category,
            "tags": params.tags,
            "language": params.language,
            "visibility": params.visibility,
            "metadata": params.metadata,
        }
    )


def _host_body(params: AddHostParams) -> dict[str, Any]:
    return _drop_none({"identity": params.identity, "role": params.role})


class LiveStreamsResource:
    def __init__(self, http: RavenHttpClient) -> None:
        self._http = http

    def create(self, params: CreateLiveStreamParams) -> LiveStream:
        result = self._http.request("/v1/live-streams", method="POST", body=_create_body(params))
        return cast(LiveStream, result)

    def list(self, status: LiveStreamStatus | None = None) -> list[LiveStream]:
        query = _drop_none({"status": status})
        return cast("list[LiveStream]", self._http.request("/v1/live-streams", query=query))

    def get(self, stream_id: str) -> LiveStream:
        """Includes the stream's live viewer count — ``list()`` does not, to avoid one SFU round trip per row."""
        return cast(LiveStream, self._http.request(f"/v1/live-streams/{stream_id}"))

    def update(self, stream_id: str, params: UpdateLiveStreamParams) -> LiveStream:
        result = self._http.request(f"/v1/live-streams/{stream_id}", method="PATCH", body=_update_body(params))
        return cast(LiveStream, result)

    def start(self, stream_id: str) -> LiveStream:
        """CREATED -> LIVE."""
        return cast(LiveStream, self._http.request(f"/v1/live-streams/{stream_id}/start", method="POST"))

    def end(self, stream_id: str) -> LiveStream:
        """LIVE -> ENDED, terminal. An ended stream cannot be restarted — create a new one."""
        return cast(LiveStream, self._http.request(f"/v1/live-streams/{stream_id}/end", method="POST"))

    def add_host(self, stream_id: str, params: AddHostParams) -> IssuedStreamCredential:
        """Registers a host/co-host and mints full-publish RTC + moderator-or-above chat credentials in one call.

        Call again with the same identity to re-mint fresh credentials.
        """
        result = self._http.request(f"/v1/live-streams/{stream_id}/hosts", method="POST", body=_host_body(params))
        return cast(IssuedStreamCredential, result)

    def remove_host(self, stream_id: str, identity: str) -> None:
        """Soft removal — the host's chat history in the stream is preserved."""
        self._http.request(f"/v1/live-streams/{stream_id}/hosts/{identity}", method="DELETE")

    def create_viewer_token(self, stream_id: str, identity: str) -> IssuedStreamCredential:
        """Always subscribe-only on RTC and MEMBER on chat — see the module doc."""
        result = self._http.request(
            f"/v1/live-streams/{stream_id}/viewer-tokens", method="POST", body={"identity": identity}
        )
        return cast(IssuedStreamCredential, result)

    def leave(self, stream_id: str, identity: str) -> None:
        """A clean-leave signal for ``live_stream.viewer_left``, not a disconnect detector.

        Raven has no way to observe an abrupt viewer disconnect in this phase.
        """
        self._http.request(f"/v1/live-streams/{stream_id}/leave", method="POST", body={"identity": identity})


class AsyncLiveStreamsResource:
    def __init__(self, http: AsyncRavenHttpClient) -> None:
        self._http = http

    async def create(self, params: CreateLiveStreamParams) -> LiveStream:
        result = await self._http.request("/v1/live-streams", method="POST", body=_create_body(params))
        return cast(LiveStream, result)

    async def list(self, status: LiveStreamStatus | None = None) -> list[LiveStream]:
        query = _drop_none({"status": status})
        return cast("list[LiveStream]", await self._http.request("/v1/live-streams", query=query))

    async def get(self, stream_id: str) -> LiveStream:
        return cast(LiveStream, await self._http.request(f"/v1/live-streams/{stream_id}"))

    async def update(self, stream_id: str, params: UpdateLiveStreamParams) -> LiveStream:
        result = await self._http.request(f"/v1/live-streams/{stream_id}", method="PATCH", body=_update_body(params))
        return cast(LiveStream, result)

    async def start(self, stream_id: str) -> LiveStream:
        return cast(LiveStream, await self._http.request(f"/v1/live-streams/{stream_id}/start", method="POST"))

    async def end(self, stream_id: str) -> LiveStream:
        return cast(LiveStream, await self._http.request(f"/v1/live-streams/{stream_id}/end", method="POST"))

    async def add_host(self, stream_id: str, params: AddHostParams) -> IssuedStreamCredential:
        result = await self._http.request(
            f"/v1/live-streams/{stream_id}/hosts", method="POST", body=_host_body(params)
        )
        return cast(IssuedStreamCredential, result)

    async def remove_host(self, stream_id: str, identity: str) -> None:
        await self._http.request(f"/v1/live-streams/{stream_id}/hosts/{identity}", method="DELETE")

    async def create_viewer_token(self, stream_id: str, identity: str) -> IssuedStreamCredential:
        result = await self._http.request(
            f"/v1/live-streams/{stream_id}/viewer-tokens", method="POST", body={"identity": identity}
        )
        return cast(IssuedStreamCredential, result)

    async def leave(self, stream_id: str, identity: str) -> None:
        await self._http.request(f"/v1/live-streams/{stream_id}/leave", method="POST", body={"identity": identity})
